From acae84750b700ee215dd21e379da0c14469cc9b5 Mon Sep 17 00:00:00 2001
Message-Id: <acae84750b700ee215dd21e379da0c14469cc9b5.1421709247.git.chang-joon.lee@intel.com>
In-Reply-To: <3650eb44734ad4fbe098313349da8185595a20d8.1421709247.git.chang-joon.lee@intel.com>
References: <3650eb44734ad4fbe098313349da8185595a20d8.1421709247.git.chang-joon.lee@intel.com>
From: Ramalingam C <ramalingam.c@intel.com>
Date: Mon, 20 Oct 2014 18:33:30 +0530
Subject: [PATCH 62/75] MUST_REBASE [VPG]: adf/intel/vlv: 180 degree HW
 rotation support

Implemented the configuration flow for the 180 degree
HW plane rotation

MUST_REBASE: Since Google ADF framework is not in upstream and also
we cannot have two display drivers, other one being i915, we need to
work on i915 and ADF convergence path before it can be upstreamed

For: GMINL-3547
Change-Id: I638baaff50ab30c743e35e5cc131866b1b888e6b
Signed-off-by: Ramalingam C <ramalingam.c@intel.com>
Signed-off-by: Deepak S <deepak.s@intel.com>
---
 drivers/video/adf/intel/core/vlv/vlv_pri_plane.c   |   15 ++++++++--
 drivers/video/adf/intel/core/vlv/vlv_sp_plane.c    |   29 ++++++++++++++++----
 .../video/adf/intel/include/core/intel_dc_config.h |   12 ++------
 .../video/adf/intel/include/core/vlv/vlv_dc_regs.h |    1 +
 4 files changed, 39 insertions(+), 18 deletions(-)

diff --git a/drivers/video/adf/intel/core/vlv/vlv_pri_plane.c b/drivers/video/adf/intel/core/vlv/vlv_pri_plane.c
index ae00cc4..82331aa 100644
--- a/drivers/video/adf/intel/core/vlv/vlv_pri_plane.c
+++ b/drivers/video/adf/intel/core/vlv/vlv_pri_plane.c
@@ -19,6 +19,7 @@
 #include <core/vlv/vlv_dc_config.h>
 #include <core/vlv/vlv_dc_regs.h>
 #include <drm/i915_drm.h>
+#include <video/intel_adf.h>
 
 #define SEC_PLANE_OFFSET 0x1000
 
@@ -104,7 +105,7 @@ static const u32 pri_supported_formats[] = {
 };
 
 static const u32 pri_supported_transforms[] = {
-	INTEL_PLANE_TRANSFORM_ROT180,
+	INTEL_ADF_TRANSFORM_ROT180,
 };
 
 static const u32 pri_supported_blendings[] = {
@@ -248,7 +249,17 @@ static int vlv_pri_calculate(struct intel_plane *plane,
 	dspaddr_offset = vlv_compute_page_offset(&src_x, &src_y,
 				buf->tiling_mode, bpp, regs->stride);
 	regs->linearoff -= dspaddr_offset;
-	regs->tileoff = (src_y << 16) | src_x;
+	if (config->transform & INTEL_ADF_TRANSFORM_ROT180) {
+		regs->dspcntr |= DISPPLANE_180_ROTATION_ENABLE;
+		regs->linearoff =  regs->linearoff + (buf->h - 1) *
+						regs->stride + buf->w * bpp;
+		regs->tileoff = (((src_y + buf->h - 1) << 16) |
+							(src_x + buf->w - 1));
+	} else {
+		regs->dspcntr &= ~DISPPLANE_180_ROTATION_ENABLE;
+		regs->tileoff = (src_y << 16) | src_x;
+	}
+
 	regs->surfaddr = (buf->gtt_offset_in_pages + dspaddr_offset);
 
 	return 0;
diff --git a/drivers/video/adf/intel/core/vlv/vlv_sp_plane.c b/drivers/video/adf/intel/core/vlv/vlv_sp_plane.c
index 4bdadfe..ecbc732 100644
--- a/drivers/video/adf/intel/core/vlv/vlv_sp_plane.c
+++ b/drivers/video/adf/intel/core/vlv/vlv_sp_plane.c
@@ -13,6 +13,7 @@
  */
 
 #include <drm/i915_drm.h>
+#include <video/intel_adf.h>
 
 #include "intel_adf_device.h"
 #include "core/common/intel_dc_regs.h"
@@ -330,23 +331,39 @@ static int vlv_sp_calculate(struct intel_plane *planeptr,
 		REG_WRITE_BITS(VLV_DDL(pipe), 0x00, mask);
 	}
 
-	if (buf->tiling_mode != I915_TILING_NONE)
-		sprctl |= SP_TILED;
-
 	sprctl |= SP_ENABLE;
 	regs->dspcntr = sprctl;
 
 	linear_offset = src_y * buf->stride + src_x * bpp;
 	sprsurf_offset = vlv_compute_page_offset(&src_x, &src_y,
 			buf->tiling_mode, bpp, buf->stride);
-	regs->linearoff = linear_offset - sprsurf_offset;
+	linear_offset -= sprsurf_offset;
 
 	regs->stride = buf->stride;
 	regs->pos = ((dst_y << 16) | dst_x);
-	regs->tileoff = (src_y << 16) | src_x;
 	regs->size = (dst_h << 16) | dst_w;
 	regs->surfaddr = (buf->gtt_offset_in_pages + sprsurf_offset);
 
+	if (buf->tiling_mode != I915_TILING_NONE) {
+		regs->dspcntr |= SP_TILED;
+		if (config->transform & INTEL_ADF_TRANSFORM_ROT180) {
+			regs->dspcntr |= DISPPLANE_180_ROTATION_ENABLE;
+			regs->tileoff = ((src_y + dst_h) << 16) |
+							(src_x + dst_w);
+		} else {
+			regs->dspcntr &= ~DISPPLANE_180_ROTATION_ENABLE;
+			regs->tileoff = (src_y << 16) | src_x;
+		}
+	} else {
+		if (config->transform & INTEL_ADF_TRANSFORM_ROT180) {
+			regs->dspcntr |= DISPPLANE_180_ROTATION_ENABLE;
+			regs->linearoff = linear_offset + dst_h *
+					regs->stride + (dst_w + 1) * bpp;
+		} else {
+			regs->dspcntr &= ~DISPPLANE_180_ROTATION_ENABLE;
+			regs->linearoff = linear_offset;
+		}
+	}
 	return 0;
 }
 
@@ -531,7 +548,7 @@ static const u32 sprite_supported_formats[] = {
 };
 
 static const u32 sprite_supported_transforms[] = {
-	INTEL_PLANE_TRANSFORM_ROT180,
+	INTEL_ADF_TRANSFORM_ROT180,
 };
 
 static const u32 sprite_supported_blendings[] = {
diff --git a/drivers/video/adf/intel/include/core/intel_dc_config.h b/drivers/video/adf/intel/include/core/intel_dc_config.h
index 8c2b340..983b26c 100644
--- a/drivers/video/adf/intel/include/core/intel_dc_config.h
+++ b/drivers/video/adf/intel/include/core/intel_dc_config.h
@@ -17,6 +17,7 @@
 
 #include <linux/kernel.h>
 #include <linux/module.h>
+#include <video/intel_adf.h>
 
 #if defined(CONFIG_ADF)
 #include <video/adf.h>
@@ -42,15 +43,6 @@ enum intel_plane_scaling {
 	INTEL_PLANE_SCALING_UPSCALING,
 };
 
-enum intel_plane_transform {
-	INTEL_PLANE_TRANSFORM_NONE,
-	INTEL_PLANE_TRANSFORM_FLIPH,
-	INTEL_PLANE_TRANSFORM_FLIPV,
-	INTEL_PLANE_TRANSFORM_ROT90,
-	INTEL_PLANE_TRANSFORM_ROT180,
-	INTEL_PLANE_TRANSFORM_ROT270,
-};
-
 enum intel_plane_compression {
 	INTEL_PLANE_DECOMPRESSION_16X4,
 };
@@ -159,7 +151,7 @@ struct intel_plane_config {
 	u8 alpha;
 	enum intel_plane_compression compression:4;
 	enum intel_plane_blending blending:4;
-	enum intel_plane_transform transform:4;
+	enum intel_adf_transform transform:4;
 	struct intel_pipe *pipe;
 };
 
diff --git a/drivers/video/adf/intel/include/core/vlv/vlv_dc_regs.h b/drivers/video/adf/intel/include/core/vlv/vlv_dc_regs.h
index 8de921a..232c2d0 100644
--- a/drivers/video/adf/intel/include/core/vlv/vlv_dc_regs.h
+++ b/drivers/video/adf/intel/include/core/vlv/vlv_dc_regs.h
@@ -3609,6 +3609,7 @@
 #define   DISPPLANE_NO_LINE_DOUBLE		0
 #define   DISPPLANE_STEREO_POLARITY_FIRST	0
 #define   DISPPLANE_STEREO_POLARITY_SECOND	(1<<18)
+#define   DISPPLANE_180_ROTATION_ENABLE		(1<<15)
 #define   DISPPLANE_TRICKLE_FEED_DISABLE	(1<<14) /* Ironlake */
 #define   DISPPLANE_TILED			(1<<10)
 #define _DSPAADDR		(VLV_DISPLAY_BASE + 0x70184)
-- 
1.7.9.5

